chore: adding kafka brokers metrics

jcountsNR opened the pull request.

jcountsNR commented:
Adding a note, this is a continuation of [this PR](https://github.com/open-telemetry/semantic-conventions/pull/113), I just could not make changes to that fork so I needed to create a new one.

jcountsNR commented:
@jsuereth , I'm tagging you since it looks like you were the assignee on this one. We have a lot of eyes on our end wanting to get this one to the finish line, please let me know if there is anything I can do to make sure this one is good to go. Thanks!

dmitryax commented on docs/messaging/kafka.md:
```diff
@@ -112,6 +112,15 @@ This section defines how to apply semantic conventions when collecting Kafka met
 | messaging.kafka.brokers                      | UpDownCounter | Int64      | brokers   | `{broker}`   | Number of brokers in the cluster. | | |
+| messaging.kafka.brokers.count                | UpDownCounter | Int64      | brokers   | `{broker}`   | sum of brokers in the cluster | | |
+| messaging.kafka.brokers.consumer.fetch.rate  | Gauge         | Double     | fetches per second | `{fetch}/s`  | Average consumer fetch Rate. | `state` | `in`, `out` |
+| messaging.kafka.brokers.network.io           | Gauge         | Double     | bytes per second   | `{By/s}`     | Average tncoming Byte Rate in bytes/second. | | |
```
This should be `Counter`
```suggestion
| messaging.kafka.brokers.network.io           | Counter         | Int64     | bytes   | `By`     |  The bytes received or sent by the broker. | | |
```
`messaging.kafka.network.io` then should be removed or deprecated

dmitryax commented on docs/messaging/kafka.md:
```diff
@@ -112,6 +112,15 @@ This section defines how to apply semantic conventions when collecting Kafka met
 | messaging.kafka.leader.elections             | Counter       | Int64      | elections | `{election}` | Leader election rate (increasing values indicates broker failures). | | |
 | messaging.kafka.leader.unclean-elections     | Counter       | Int64      | elections | `{election}` | Unclean leader election rate (increasing values indicates broker failures). | | |
 | messaging.kafka.brokers                      | UpDownCounter | Int64      | brokers   | `{broker}`   | Number of brokers in the cluster. | | |
```
This metric now should be removed/deprecated

jcountsNR replied:
Done, I moved to a new section called `### Broker Metrics` and I removed the old metric as well.

dmitryax reviewed (commented):
Please move the metrics to another section `### Broker Metrics`

lmolkova commented on docs/messaging/kafka.md:
```diff
@@ -159,4 +158,20 @@ This section defines how to apply semantic conventions when collecting Kafka met
+| messaging.kafka.brokers.count                | UpDownCounter | Int64      | brokers   | `{broker}`   | sum of brokers in the cluster | | |
+| messaging.kafka.brokers.consumer.fetch.rate  | Gauge         | Double     | fetches per second | `{fetch}/s`  | Average consumer fetch Rate. | `state` | `in`, `out` |
+| messaging.kafka.brokers.network.io           | Counter       | Int64      | bytes   | `By`     |  The bytes received or sent by the broker. | | |
+| messaging.kafka.brokers.requests.latency     | Gauge         | Double     | ms        | `{ms}` | Average Request latency in ms. | | |
```
is it possible to report it as a histogram? Then we'll have percentiles and `messaging.kafka.brokers.requests.rate` can also be derived from it?

jcountsNR replied:
Hmm, I don't have any experience converting histograms from metrics.go into histograms for OTeL. Are there other examples where someone has done this in the past? The kafka consumers for example is already tracking `lag` as a `gauge` metric, so I thought the same would make sense here.

I could see it potentially being an upgrade, I'm just not sure how to apply it to update the collector.

lmolkova commented on docs/messaging/kafka.md:
```diff
@@ -159,4 +158,20 @@ This section defines how to apply semantic conventions when collecting Kafka met
+
+| Name                                         | Instrument    | Value type | Unit   | Unit ([UCUM](/docs/general/metrics.md#instrument-units)) | Description    | Attribute Key | Attribute Values |
+| ---------------------------------------------| ------------- | ---------- | ------ | -------------------------------------------- | -------------- | ------------- | ---------------- |
+| messaging.kafka.brokers.count                | UpDownCounter | Int64      | brokers   | `{broker}`   | sum of brokers in the cluster | | |
```
Is this metric necessary?

assuming brokers report unique instance id (e.g. standard `service.instance.id` attribute), it can be derived from other metrics in this list. If brokers also report standard metrics (CPU, memory, etc), this can also be derived from them.

jcountsNR replied:
That might be true, this is primarily here to replace the `kafka.brokers` metric which already exists. So doing away with this completely might impact metrics that are being collected now, although I do see the point that it could be derived if we added the attribute.

lmolkova commented on docs/messaging/kafka.md:
```diff
@@ -159,4 +158,20 @@ This section defines how to apply semantic conventions when collecting Kafka met
+| Name                                         | Instrument    | Value type | Unit   | Unit ([UCUM](/docs/general/metrics.md#instrument-units)) | Description    | Attribute Key | Attribute Values |
+| ---------------------------------------------| ------------- | ---------- | ------ | -------------------------------------------- | -------------- | ------------- | ---------------- |
+| messaging.kafka.brokers.count                | UpDownCounter | Int64      | brokers   | `{broker}`   | sum of brokers in the cluster | | |
+| messaging.kafka.brokers.consumer.fetch.rate  | Gauge         | Double     | fetches per second | `{fetch}/s`  | Average consumer fetch Rate. | `state` | `in`, `out` |
```
If brokers can report `messaging.kafka.brokers.consumer.fetch.count`, it can provide more information and the rate can be derived from it - can it be changed?

AndrewJSchofield replied:
Indeed. I thought that it was preferred not to have rate metrics but instead use a counter from which the rate can be derived for whatever time period is desired.

jcountsNR replied:
Hmm, yeah I think that is correct. This should be updated.

lmolkova commented on docs/messaging/kafka.md:
```diff
@@ -159,4 +158,20 @@ This section defines how to apply semantic conventions when collecting Kafka met
+| messaging.kafka.brokers.consumer.fetch.rate  | Gauge         | Double     | fetches per second | `{fetch}/s`  | Average consumer fetch Rate. | `state` | `in`, `out` |
+| messaging.kafka.brokers.network.io           | Counter       | Int64      | bytes   | `By`     |  The bytes received or sent by the broker. | | |
+| messaging.kafka.brokers.requests.latency     | Gauge         | Double     | ms        | `{ms}` | Average Request latency in ms. | | |
+| messaging.kafka.brokers.requests.rate        | Gauge         | Double     | requests per second | `{request}/s`| Average request rate per second. | | |
```
same here, `messaging.kafka.brokers.request.count` is more flexible and the rate can be derived from it.

jcountsNR replied:
agreed.

lmolkova commented on docs/messaging/kafka.md:
```diff
@@ -159,4 +158,20 @@ This section defines how to apply semantic conventions when collecting Kafka met
+| messaging.kafka.brokers.network.io           | Counter       | Int64      | bytes   | `By`     |  The bytes received or sent by the broker. | | |
+| messaging.kafka.brokers.requests.latency     | Gauge         | Double     | ms        | `{ms}` | Average Request latency in ms. | | |
+| messaging.kafka.brokers.requests.rate        | Gauge         | Double     | requests per second | `{request}/s`| Average request rate per second. | | |
+| messaging.kafka.brokers.requsts.size         | Gauge         | Double     | bytes     | `By`         | Average request size in bytes. | | |
```
this would probably be best represented with histogram to allow distribution and then a rate counter is not needed.
Another possibility would be to only report `messaging.kafka.brokers.network.io` with direction attribute or use  `messaging.kafka.brokers.request.bytes` counter counting all the bytes .

jcountsNR replied:
Hi @lmolkova , is there an example of a histogram metric available and how that would be defined?

lmolkova replied:
you can find examples throughout this repo, for example, take a look at `http.client.request.duration`. there is a lot of information in the [spec repo](https://github.com/open-telemetry/opentelemetry-specification/blob/main/specification/metrics) and also take a look at metric docs on [opentelemetry.io](https://opentelemetry.io/docs/specs/otel/metrics/data-model/#histogram)

lmolkova commented on docs/messaging/kafka.md:
```diff
@@ -159,4 +158,20 @@ This section defines how to apply semantic conventions when collecting Kafka met
+| messaging.kafka.brokers.requests.latency     | Gauge         | Double     | ms        | `{ms}` | Average Request latency in ms. | | |
+| messaging.kafka.brokers.requests.rate        | Gauge         | Double     | requests per second | `{request}/s`| Average request rate per second. | | |
+| messaging.kafka.brokers.requsts.size         | Gauge         | Double     | bytes     | `By`         | Average request size in bytes. | | |
+| messaging.kafka.brokers.responses.rate       | Gauge         | Double     | responses per second| `{response}/s`| Average response rate per second. | | |
```
same comment as on request rate

lmolkova commented on docs/messaging/kafka.md:
```diff
@@ -159,4 +158,20 @@ This section defines how to apply semantic conventions when collecting Kafka met
+| messaging.kafka.brokers.requests.rate        | Gauge         | Double     | requests per second | `{request}/s`| Average request rate per second. | | |
+| messaging.kafka.brokers.requsts.size         | Gauge         | Double     | bytes     | `By`         | Average request size in bytes. | | |
+| messaging.kafka.brokers.responses.rate       | Gauge         | Double     | responses per second| `{response}/s`| Average response rate per second. | | |
+| messaging.kafka.brokers.response_size        | Gauge         | Double     | bytes     | `By`         | Average response size in bytes. | | |
```
same comment as on request size

lmolkova commented on docs/messaging/kafka.md:
```diff
@@ -159,4 +158,20 @@ This section defines how to apply semantic conventions when collecting Kafka met
+| messaging.kafka.brokers.requsts.size         | Gauge         | Double     | bytes     | `By`         | Average request size in bytes. | | |
+| messaging.kafka.brokers.responses.rate       | Gauge         | Double     | responses per second| `{response}/s`| Average response rate per second. | | |
+| messaging.kafka.brokers.response_size        | Gauge         | Double     | bytes     | `By`         | Average response size in bytes. | | |
+| messaging.kafka.brokers.requests.in.flight   | Gauge         | Int64      | requests  | `{request}`  | Requests in flight. | | |
```
maybe `messaging.kafka.brokers.active_requests` to stay consistent with HTTP semantic conventions.

lmolkova reviewed (changes requested):
I'm not sure if this PR documents what's already done on Kafka and there are limitations on what can be changed, but I suggest to stay as close to general otel metric requirements as possible.
I left several comments on this.

Also, I'd recommend using `messaging.kafka.broker` (not `brokers`) as a namespace.

lmolkova commented on docs/messaging/kafka.md:
```diff
@@ -159,4 +158,20 @@ This section defines how to apply semantic conventions when collecting Kafka met
+
+**Description:** Kafka Broker level metrics.
+
+| Name                                         | Instrument    | Value type | Unit   | Unit ([UCUM](/docs/general/metrics.md#instrument-units)) | Description    | Attribute Key | Attribute Values |
```
I don't see any attributes - are there none? We should have at least standard [`service.*` ones](https://github.com/open-telemetry/semantic-conventions/blob/main/docs/resource/README.md#service)

AndrewJSchofield replied:
I would suggest using the broker's id (which I would call "node id" instead of "broker id" following Kafka best practice).

AndrewJSchofield commented on docs/messaging/kafka.md:
```diff
@@ -159,4 +158,20 @@ This section defines how to apply semantic conventions when collecting Kafka met
+| ---------------------------------------------| ------------- | ---------- | ------ | -------------------------------------------- | -------------- | ------------- | ---------------- |
+| messaging.kafka.brokers.count                | UpDownCounter | Int64      | brokers   | `{broker}`   | sum of brokers in the cluster | | |
+| messaging.kafka.brokers.consumer.fetch.rate  | Gauge         | Double     | fetches per second | `{fetch}/s`  | Average consumer fetch Rate. | `state` | `in`, `out` |
+| messaging.kafka.brokers.network.io           | Counter       | Int64      | bytes   | `By`     |  The bytes received or sent by the broker. | | |
```
Wouldn't we want to measure bytes sent and bytes received separately?

jcountsNR replied:
That was the original thought, but it was changed. I think changing it back makes sense, the PR I have for the kafka collector is separated.

jcountsNR commented:
Hi @AndrewJSchofield @lmolkova @dmitryax . I made some updates and raised the commit. It looks correct to me now, let me know if I've missed anything.

jcountsNR commented:
One other follow up, I don't think that node and broker are interchangeable, and since this is specific to broker metrics I'm inclined to say we should leave it as broker.id. I did add the attribute to everything except broker count.

jcountsNR commented:
@dmitryax , I  updated a few things based on the semantic-convention conversation. I've run all of the unit tests except the integration_test which I haven't been able to get to run locally. I re-generated everything and that didn't fail, and neither did the lint tool. here is what I updated:

1. changed the `rate` into `count` for all of the metrics except the incoming and outgoing byte rate
2. updated some of the value types
3. Changed the new count metrics to sums instead of gauges
4. Renamed them with the `broker.count` convention instead of the `brokers.count`

There were a lot of changes but hopefully the workflow will pass. The only thing I didn't update was the suggestion to convert two of the fields into histograms. I saw that there were multiple similar metrics like `kafka.consumer_group.members` which are still gauge types, so I think it's consistent still with what's there.

jcountsNR commented:
Sorry for a delayed response to this. This is for a very specific integration which uses sarama metrics, and it has some limitations. Since that is the case, I'm not sure that making significant updates to these metrics makes sense, because they aren't possible with the one integration so far that will be using them.

The broker related metrics for example don't use `count`, but rather `rate` for all of the metrics that they are translating for kafka. I think this PR should probably either be cancelled or updated to be a more general naming translation for the metrics we can get from confluent cloud brokers.

pyohannes commented:
In a semantic conventions SIG meeting an agreement was reached to _remove Kafka broker metrics from the semantic conventions_. See #338, which removes the PR and gives a list of reasons.

@jcountsNR, if you have any opinions please weigh in on #338. Otherwise let's close this PR when/if #338 is merged.

jcountsNR commented:
@pyohannes I'm good to close it, I don't think it's valid anymore.
